### DISCUSSION TITLE
I wanner use the stdin ,stdout and tcp at the sametime

### DISCUSSION THREAD

**Discussion by Decodetalkers** [ebkalderon/tower-lsp#342] [Q&A] [answered]
Use stdin and stdout, the editor will work with the lsp , and Sometime I cannot view the log easily, I wanner use tcp socket to view the logs ,so if it is posible with tower-lsp?

**Comment by silvanshade** [comment, marked as answer]
Which logs are you referring to? Do you mean a trace of the LSP messages between the client and server?

If you have created a vscode extension for the language server, you can view the LSP message trace in the vscode editor by enabling a specific [trace setting](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide#logging-support-for-language-server).

Otherwise, I believe you could possibly do this with tower middleware but I don't have any particular pointers for where to start with that. You might try asking on the tokio discord in the tower channel.

In theory you should also be able to do this by implementing custom types `I: AsyncRead` and `O: AsyncWrite` for the [Server](https://docs.rs/tower-lsp/latest/tower_lsp/struct.Server.html#impl) impl which you pass to `Server::new`.

In that case, you would have some custom type `I: AsyncRead` that behaves like stdin but also tees the output to a tcp socket (and similarly for `O: AsyncWrite`) that you can attach, or something like that. But implementing something like that is not basic functionality and I'm not sure if any exiting crates make that easy for you either; you might have to create a custom wrapper.
